plugins/phpfpm/phpfpm.go

```diff
@@ -0,0 +1,187 @@
+	PF_IDLE_PROCESSES       = "idle processes"
+	PF_ACTIVE_PROCESSES     = "active processes"
+	PF_TOTAL_PROCESSES      = "total processes"
+	PF_MAX_ACTIVE_PROCESSES = "max active processes"
+	PF_MAX_CHILDREN_REACHED = "max children reached"
+	PF_SLOW_REQUESTS        = "slow requests"
+)
+
+type metric map[string]int64
+type poolStat map[string]metric
+
+type phpfpm struct {
+	Statusurls []string
+
+	client *http.Client
+}
+
+var sampleConfig = `
+	# An array of address to gather stats about. Specify an ip on hostname
+	# with optional port and path. ie localhost, 10.10.3.33/server-status, etc.
+	#
+	# We can configure int two modes:
+	#   - unixsocket: the string is the path to fpm socket like /var/run/php5-fpm.sock
+	#   - http: the URL has to start with http:// or https://
+	#
+	# If no servers are specified, then default to 127.0.0.1/server-status
+	statusurls = ["http://localhost/status"]
+	# statusurls = ["http://localhost/status?full"]
+	#
+	# Or you can also use local socket.
+	# statusurls = ["192.168.10.10:/var/run/php5-fpm-www2.sock", "192.168.10.10:/var/run/php5-fpm-www1.sock"]
+	# in socket mode, you have to specifiy IP address of the host in syntax "IP:socket_path"
+	#
+	# Or using both of socket and http status
+	# statusurls = ["http://noty.ax/status", "192.168.10.10:/var/run/php5-fpm-www1.sock"]
+`
+
+func (r *phpfpm) SampleConfig() string {
+	return sampleConfig
+}
+
+func (r *phpfpm) Description() string {
+	return "Read metrics of phpfpm, via HTTP status page or socket(pending)"
+}
+
+// Reads stats from all configured servers accumulates stats.
+// Returns one of the errors encountered while gather stats (if any).
+func (g *phpfpm) Gather(acc plugins.Accumulator) error {
+	if len(g.Statusurls) == 0 {
+		return g.gatherServer("http://127.0.0.1/status", acc)
+	}
+
+	var wg sync.WaitGroup
+
+	var outerr error
+
+	for _, serv := range g.Statusurls {
+		wg.Add(1)
+		go func(serv string) {
+			defer wg.Done()
+			outerr = g.gatherServer(serv, acc)
+		}(serv)
+	}
+
+	wg.Wait()
+
+	return outerr
+}
+
+// Request status page to get stat raw data
+func (g *phpfpm) gatherServer(addr string, acc plugins.Accumulator) error {
+	if g.client == nil {
+
+		client := &http.Client{}
+		g.client = client
+	}
+
+	if strings.HasPrefix(addr, "http://") || strings.HasPrefix(addr, "https://") {
+		u, err := url.Parse(addr)
+		if err != nil {
+			return fmt.Errorf("Unable parse server address '%s': %s", addr, err)
+		}
+
+		req, err := http.NewRequest("GET", fmt.Sprintf("%s://%s%s", u.Scheme, u.Host, u.Path), nil)
+		res, err := g.client.Do(req)
+		if err != nil {
+			return fmt.Errorf("Unable to connect to phpfpm status page '%s': %v", addr, err)
+		}
+
+		if res.StatusCode != 200 {
+			return fmt.Errorf("Unable to get valid stat result from '%s': %v", addr, err)
+		}
+
+		importMetric(res.Body, acc, u.Host)
+	} else {
+		socketAddr := strings.Split(addr, ":")
+
+		fcgi, _ := NewClient("unix", socketAddr[1])
```

Comment:
Here, we are requesting via FastCGI, using the socket liket `/var/run/php5-fpm.sock`